### PR TITLE
=per #17959 PersistentFsmActor persists StateChangeEvent on stay

### PR DESCRIPTION
**Pull request by leonidb**
Prevent PersistentFSM from saving state change event on stay

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
PLS BUILD

File: akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala [hunk 1]
```diff
@@ -17,9 +17,12 @@ import scala.reflect.ClassTag
  * FSM actor implementation with persistent state
  *
  * Supports the usual [[akka.actor.FSM]] functionality with additional persistence features.
- * State and State Data are persisted on every state change.
- * FSM is identified by 'persistenceId' value.
+ * State changes are persisted. FSM is identified by 'persistenceId' value.
  * Persistence execution order is: persist -&gt; wait for ack -&gt; apply state. Incoming messages are deferred until the state is applied.
+ *
+ * Persistent FSM DSL provides a possibility to persist domain events, which are saved atomically with the State change event,
+ * meaning they either both peresited successfully, or both fail.
```
Comment (ahjohannessen) [inline akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala]: *peresited* -> persisted :)

File: akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala [hunk 2]
```diff
@@ -17,9 +17,12 @@ import scala.reflect.ClassTag
...
+ * State Data is constructed based on domain events, according to user's implementatio of applyEvent function.
```
Comment (ahjohannessen) [inline akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala]: *implementatio* -> implementation

**Comment by leonidb** [comment]
Thanks, @ahjohannessen, 
it was a middle-of-the-night commit :)

**Comment by leonidb** [comment]
@ktoso 
Any comments on this one, before I start rebasing and squashing?

File: akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala [hunk 3]
```diff
@@ -17,9 +17,12 @@ import scala.reflect.ClassTag
  * FSM actor implementation with persistent state
  *
  * Supports the usual [[akka.actor.FSM]] functionality with additional persistence features.
- * State and State Data are persisted on every state change.
- * FSM is identified by 'persistenceId' value.
+ * State changes are persisted. FSM is identified by 'persistenceId' value.
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala]: Maybe: `State changes are persisted on every state change (except when `stay` is used)`

File: akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala [hunk 4]
```diff
@@ -17,9 +17,12 @@ import scala.reflect.ClassTag
...
  * Persistence execution order is: persist -&gt; wait for ack -&gt; apply state. Incoming messages are deferred until the state is applied.
+ *
+ * Persistent FSM DSL provides a possibility to persist domain events, which are saved atomically with the State change event,
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala]: remove `DSL` please

File: akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala (hunk 3 above)
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala]: Ok, let's do:

- Remove the section https://github.com/akka/akka/pull/17974/files#diff-ef6f00459ea008459f3774f7875a3e5eR23
- Make the above: "State changes are persisted atomically together with domain events, which means that either both succeed or both fail, i.e. a state transition event will not be stored if persistence of an event related to that change fails".

Is that right?

Comment (leonidb) [inline akka-persistence/src/main/scala/akka/persistence/fsm/PersistentFSMActor.scala]: Sounds good.

File: akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMActorSpec.scala [hunk 5]
```diff
@@ -223,6 +223,56 @@ abstract class PersistentFSMActorSpec(config: Config) extends PersistenceSpec(co
...
+      expectMsg(NonEmptyShoppingCart(List(shirt, shoes, coat)))
+
+      expectMsg(NonEmptyShoppingCart(List(shirt, shoes, coat)))
+
+      expectTerminated(fsmRef)
+
+      val persistentEventsStreamer = system.actorOf(PersistentEventsStreamer.props(persistenceId, testActor))
+
+      expectMsg(ItemAdded(Item("1", "Shirt", 59.99F)))
+      expectMsgClass(classOf[StateChangeEvent])   //because a timeout is defined, State Change is persisted
```
Comment (ktoso) [inline akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMActorSpec.scala]: minor, `expectMsgType[StateChangeEvent]` is less typing :)

File: akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMActorSpec.scala [hunk 6]
```diff
@@ -349,6 +399,20 @@ object PersistentFSMActorSpec {
     def props(persistenceId: String, reportActor: ActorRef) =
       Props(new WebStoreCustomerFSMActor(persistenceId, reportActor))
   }
+
+  class PersistentEventsStreamer(id: String, client: ActorRef) extends PersistentView {
```
Comment (ktoso) [inline akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMActorSpec.scala]: Let's make it a `PersistentActor`, as we'll try to remove `PersistentView` soon.

File: akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMActorSpec.scala (hunk 5 above)
Comment (leonidb) [inline akka-persistence/src/test/scala/akka/persistence/fsm/PersistentFSMActorSpec.scala]: I'll fix it

**Comment by leonidb** [comment]
@ktoso 
What about the implementation itself? Is it OK? 
```def applyState(nextState: State): Unit``` got pretty complicated...

I've tried to use ```deferAsync``` as a handler that is called after all event handlers were called, but it doesn't defer the incoming events. Would be handy to have a "sync" version of ```deferAsync``` or some other special handler called when all handlers were executed.

As I haven't found anything like the above, I've used a counter.

**Comment by patriknw** [comment]
LGTM, I'm fine with the counter.
Please rebase. We are releasing 2.3-M3 on Friday and it would be great to include this.
cc @ktoso

**Comment by patriknw** [comment]
please also squash into one commit

**Comment by leonidb** [comment]
@patriknw 
Great! I'll try to finish it today.

**Comment by leonidb** [comment]
@ktoso 
The merge with your changes led to a test failure. I think that this failure was due to a small bug that existed in PersistentFSM, and the current behavior is actually the correct one. See the comments to my last commit.
Please take a look, I'll squash the last 2 commits if everything is OK.

**Comment by leonidb** [comment]
@ktoso @patriknw 
After giving it some thought, I think I was wrong. In the regular FSM every command handled by the actor resets the timeout, and then the new state either sets the timeout again or leaves it as is.

So, rewriting the explanation of the test failure from my last commit:
1. ```AddItem``` is handled -> ```StateChangeEvent``` is persisted, setting a timeout to 1 sec
2. ```GetCurrentCart``` is handled -> ```StateChangeEvent``` is not persisted since the event doesn't  cause a state to change, doesn't emit any domain events and doesn't set the timeout to anything.
 In the usual ```FSM``` flow not setting timeout resets the previously set timeout.
 In ```PersistentFSM``` prior to this bug fix a None timeout was persisted, leading to recovery without state timeout.
After the bug fix ```PersistentFSM``` doesn't store the ```StateChangeEvent```, leading to recovery with the timeout of 1 sec.

Can be fixed by storing the last applied timeout  in a ```var``` in ```PersistentFSM``` and using it to determine if the new state changes the previously set timeout.

**Comment by leonidb** [comment]
@ktoso @patriknw 
Alternatively, this can be considered a valid behavior, since in ```PersistentFSM``` events that don't affect domain and don't change state shouldn't normally reset receive timeout?

This does feel like a corner case which should rarely happen in real ```PersistentFSM``` implementation.

Maybe this can be addressed later, as  a part of the effort to remove the duplication between ```FSM``` and ```PersistentFSM```?

Not sure if this is worth fixing right now.
What do you think?

**Comment by patriknw** [comment]
I'm not sure what semantics that are desired, but I'm fine with leaving it as a corner case and adjust in a separate PR, if needed. Please add a separate issue, or write a comment in the DRY PersistentFSM/FSM issue.

**Comment by patriknw** [comment]
PLS BUILD

**Comment by patriknw** [comment]
@ktoso this is ready right, I'd like to merge before we introduce any new merge conflict

**Comment by ktoso** [comment]
LGTM, let's open a ticket to investigate exact semantics of timeouts in FSM - would you open the ticket @leonidb? Thanks!

**Comment by patriknw** [comment]
great, let's merge after final pr validation

**Comment by leonidb** [comment]
> investigate exact semantics of timeouts in FSM

@ktoso Do you mean ```PersistentFSM``` timeout following recovery issue that has led to the test failure, or a more general investigation of ```FSM``` state timeouts?

**Comment by ktoso** [comment]
I mean what you described in comments https://github.com/akka/akka/pull/17974#issuecomment-130554047 https://github.com/akka/akka/pull/17974#issuecomment-130559219 I'm not sure yet myself which behaviour is valid, will need to focus on it, so a issue would help :)

**Comment by patriknw** [comment]
Refs #17959

**Pull request merged**